File: julia/src/benchmarks/Maze.jl
@enum CellKind begin
    WALL = 0
    SPACE = 1
    START = 2
    FINISH = 3
    BORDER = 4
    PATH = 5
end

is_walkable(kind::CellKind) = kind in (SPACE, START, FINISH)

mutable struct Cell
    kind::CellKind
    neighbors::Vector{Cell}
    x::Int
    y::Int
end

function Cell(x::Int, y::Int)
    return Cell(WALL, Cell[], x, y)
end

function add_neighbor(cell::Cell, neighbor::Cell)
    push!(cell.neighbors, neighbor)
end

function reset(cell::Cell)
    if cell.kind == SPACE
        cell.kind = WALL
    end
end

mutable struct Maze
    width::Int
    height::Int
    cells::Matrix{Cell}
    start::Cell
    finish::Cell
end

function Maze(width::Int, height::Int)
    w = max(width, 5)
    h = max(height, 5)

    cells = Matrix{Cell}(undef, h, w)
    for y = 1:h
        for x = 1:w
            cells[y, x] = Cell(x-1, y-1)
        end
    end

    start_cell = cells[2, 2]
    finish_cell = cells[h-1, w-1]
    start_cell.kind = START
    finish_cell.kind = FINISH

    maze = Maze(w, h, cells, start_cell, finish_cell)
    update_neighbors(maze)
    return maze
end

function update_neighbors(maze::Maze)

    for y = 1:maze.height
        for x = 1:maze.width
            empty!(maze.cells[y, x].neighbors)
        end
    end

    for y = 1:maze.height
        for x = 1:maze.width
            cell = maze.cells[y, x]

            if x > 1 && y > 1 && x < maze.width && y < maze.height
                add_neighbor(cell, maze.cells[y-1, x])
                add_neighbor(cell, maze.cells[y+1, x])
                add_neighbor(cell, maze.cells[y, x+1])
                add_neighbor(cell, maze.cells[y, x-1])

                for _ = 1:4
                    i = Helper.next_int(4) + 1
                    j = Helper.next_int(4) + 1
                    if i != j && i <= length(cell.neighbors) && j <= length(cell.neighbors)
                        cell.neighbors[i], cell.neighbors[j] =
                            cell.neighbors[j], cell.neighbors[i]
                    end
                end
            else
                cell.kind = BORDER
            end
        end
    end
end

function reset(maze::Maze)
    for y = 1:maze.height
        for x = 1:maze.width
            reset(maze.cells[y, x])
        end
    end
    maze.start.kind = START
    maze.finish.kind = FINISH
end

function dig(maze::Maze, start_cell::Cell)
    stack = Cell[]
    push!(stack, start_cell)

    while !isempty(stack)
        cell = pop!(stack)

        walkable = 0
        for n in cell.neighbors
            if is_walkable(n.kind)
                walkable += 1
            end
        end

        if walkable != 1
            continue
        end

        cell.kind = SPACE

        for n in cell.neighbors
            if n.kind == WALL
                push!(stack, n)
            end
        end
    end
end

function ensure_open_finish(maze::Maze, start_cell::Cell)
    stack = Cell[]
    push!(stack, start_cell)

    while !isempty(stack)
        cell = pop!(stack)

        cell.kind = SPACE

        walkable = 0
        for n in cell.neighbors
            if is_walkable(n.kind)
                walkable += 1
            end
        end

        if walkable > 1
            continue
        end

        for n in cell.neighbors
            if n.kind == WALL
                push!(stack, n)
            end
        end
    end
end

function generate(maze::Maze)
    for n in maze.start.neighbors
        if n.kind == WALL
            dig(maze, n)
        end
    end

    for n in maze.finish.neighbors
        if n.kind == WALL
            ensure_open_finish(maze, n)
        end
    end
end

function middle_cell(maze::Maze)::Cell
    return maze.cells[div(maze.height, 2)+1, div(maze.width, 2)+1]
end

function checksum(maze::Maze)::UInt32
    hasher = UInt32(2166136261)
    prime = UInt32(16777619)

    for y = 1:maze.height
        for x = 1:maze.width
            if maze.cells[y, x].kind == SPACE
                val = UInt32((x-1) * (y-1))
                hasher = (hasher ⊻ val) * prime
            end
        end
    end
    return hasher
end

function print_to_console(maze::Maze)
    for y = 1:maze.height
        for x = 1:maze.width
            cell = maze.cells[y, x]
            if cell.kind == SPACE
                print(" ")
            elseif cell.kind == WALL
                print("\u001B[34m#\u001B[0m")
            elseif cell.kind == BORDER
                print("\u001B[31mO\u001B[0m")
            elseif cell.kind == START
                print("\u001B[32m>\u001B[0m")
            elseif cell.kind == FINISH
                print("\u001B[32m<\u001B[0m")
            elseif cell.kind == PATH
                print("\u001B[33m.\u001B[0m")
            end
        end
        println()
    end
    println()
end

mutable struct MazeGenerator <: AbstractBenchmark
    width::Int64
    height::Int64
    maze::Union{Nothing,Maze}
    result_val::UInt32
end

function MazeGenerator()
    width = Helper.config_i64("Maze::Generator", "w")
    height = Helper.config_i64("Maze::Generator", "h")
    return MazeGenerator(width, height, nothing, UInt32(0))
end

name(b::MazeGenerator)::String = "Maze::Generator"

function prepare(b::MazeGenerator)
    b.maze = Maze(b.width, b.height)
    b.result_val = UInt32(0)
end

function run(b::MazeGenerator, iteration_id::Int64)
    if b.maze === nothing
        return
    end
    reset(b.maze)
    generate(b.maze)
    b.result_val += UInt32(middle_cell(b.maze).kind)
end

function checksum(b::MazeGenerator)::UInt32
    if b.maze === nothing
        return UInt32(0)
    end
    return b.result_val + checksum(b.maze)
end

struct BfsPathNode
    cell::Cell
    parent::Int
end

using DataStructures: Queue, enqueue!, dequeue!

mutable struct MazeBFS <: AbstractBenchmark
    width::Int64
    height::Int64
    maze::Union{Nothing,Maze}
    result_val::UInt32
    path::Vector{Cell}
end

function MazeBFS()
    width = Helper.config_i64("Maze::BFS", "w")
    height = Helper.config_i64("Maze::BFS", "h")
    return MazeBFS(width, height, nothing, UInt32(0), Cell[])
end

name(b::MazeBFS)::String = "Maze::BFS"

function prepare(b::MazeBFS)
    b.maze = Maze(b.width, b.height)
    generate(b.maze)
    b.result_val = UInt32(0)
    b.path = Cell[]
end

function bfs(maze::Maze, start::Cell, target::Cell)::Vector{Cell}
    if start === target
        return [start]
    end

    queue = Queue{Int}()
    visited = [falses(maze.width) for _ = 1:maze.height]
    path_nodes = BfsPathNode[]

    visited[start.y+1][start.x+1] = true
    push!(path_nodes, BfsPathNode(start, -1))
    enqueue!(queue, 1)

    while !isempty(queue)
        path_id = dequeue!(queue)
        node = path_nodes[path_id]

        for neighbor in node.cell.neighbors
            if neighbor === target
                result = [target]
                cur = path_id
                while cur > 0
                    push!(result, path_nodes[cur].cell)
                    cur = path_nodes[cur].parent
                end
                return reverse(result)
            end

            if is_walkable(neighbor.kind) && !visited[neighbor.y+1][neighbor.x+1]
                visited[neighbor.y+1][neighbor.x+1] = true
                push!(path_nodes, BfsPathNode(neighbor, path_id))
                enqueue!(queue, length(path_nodes))
            end
        end
    end

    return Cell[]
end

function mid_cell_checksum(path::Vector{Cell})::UInt32
    if isempty(path)
        return UInt32(0)
    end
    cell = path[div(length(path), 2)+1]
    return UInt32(cell.x * cell.y)
end

function run(b::MazeBFS, iteration_id::Int64)
    if b.maze === nothing
        return
    end
    b.path = bfs(b.maze, b.maze.start, b.maze.finish)
    b.result_val += UInt32(length(b.path))
end

function checksum(b::MazeBFS)::UInt32
    return b.result_val + mid_cell_checksum(b.path)
end

using DataStructures: BinaryMinHeap

mutable struct MazeAStar <: AbstractBenchmark
    width::Int64
    height::Int64
    maze::Union{Nothing,Maze}
    result_val::UInt32
    path::Vector{Cell}
end

function MazeAStar()
    width = Helper.config_i64("Maze::AStar", "w")
    height = Helper.config_i64("Maze::AStar", "h")
    return MazeAStar(width, height, nothing, UInt32(0), Cell[])
end

name(b::MazeAStar)::String = "Maze::AStar"

function prepare(b::MazeAStar)
    b.maze = Maze(b.width, b.height)
    generate(b.maze)
    b.result_val = UInt32(0)
    b.path = Cell[]
end

function heuristic(a::Cell, b::Cell)::Int32
    return Int32(abs(a.x - b.x) + abs(a.y - b.y))
end

function idx(y::Int, x::Int, width::Int)::Int64
    return Int64((y) * width + x)
end

function astar(maze::Maze, start::Cell, target::Cell)::Vector{Cell}
    if start === target
        return [start]
    end

    width = maze.width
    height = maze.height
    size = width * height

    came_from = fill(-1, size)
    g_score = fill(typemax(Int32), size)
    best_f = fill(typemax(Int32), size)

    start_idx = start.y * width + start.x + 1
    target_idx = target.y * width + target.x + 1

    open_set = BinaryMinHeap{Tuple{Int32,Int64}}()
    in_open = falses(size)

    g_score[start_idx] = 0
    f_start = heuristic(start, target)
    push!(open_set, (f_start, start_idx))
    best_f[start_idx] = f_start
    in_open[start_idx] = true

    while !isempty(open_set)
        f_val, current_idx = pop!(open_set)
        in_open[current_idx] = false

        if f_val != best_f[current_idx]
            continue
        end

        if current_idx == target_idx
            result = Cell[]
            cur = current_idx
            while cur != -1
                y = div(cur - 1, width) + 1
                x = (cur - 1) % width + 1
                push!(result, maze.cells[y, x])
                cur = came_from[cur]
            end
            return reverse(result)
        end

        current_y = div(current_idx - 1, width) + 1
        current_x = (current_idx - 1) % width + 1
        current_cell = maze.cells[current_y, current_x]
        current_g = g_score[current_idx]

        for neighbor in current_cell.neighbors
            if !is_walkable(neighbor.kind)
                continue
            end

            neighbor_idx = neighbor.y * width + neighbor.x + 1
            tentative_g = Int32(current_g + 1)

            if tentative_g < g_score[neighbor_idx]
                came_from[neighbor_idx] = current_idx
                g_score[neighbor_idx] = tentative_g
                f_new = tentative_g + heuristic(neighbor, target)

                if f_new < best_f[neighbor_idx]
                    best_f[neighbor_idx] = f_new
                    if in_open[neighbor_idx]

                        push!(open_set, (f_new, neighbor_idx))
                    else
                        push!(open_set, (f_new, neighbor_idx))
                        in_open[neighbor_idx] = true
                    end
                end
            end
        end
    end

    return Cell[]
end

function mid_cell_checksum(path::Vector{Cell})::UInt32
    if isempty(path)
        return UInt32(0)
    end
    cell = path[div(length(path), 2)+1]
    return UInt32(cell.x * cell.y)
end

function run(b::MazeAStar, iteration_id::Int64)
    if b.maze === nothing
        return
    end
    b.path = astar(b.maze, b.maze.start, b.maze.finish)
    b.result_val += UInt32(length(b.path))
end

function checksum(b::MazeAStar)::UInt32
    return b.result_val + mid_cell_checksum(b.path)
end

File: nim/src/benchmarks/maze.nim
import std/[math, algorithm, deques, tables, heapqueue]
import ../benchmark
import ../helper

type
  CellKind* = enum
    Wall = 0
    Space = 1
    Start = 2
    Finish = 3
    Border = 4
    Path = 5

  Cell* = ref object
    kind*: CellKind
    neighbors*: seq[Cell]
    x*, y*: int

  Maze* = ref object
    width*, height*: int
    cells*: seq[seq[Cell]]
    start*, finish*: Cell

  MazeGenerator* = ref object of Benchmark
    resultVal: uint32
    width, height: int32
    maze: Maze

  PathNode* = object
    cell: Cell
    parent: int

  MazeBFS* = ref object of Benchmark
    resultVal: uint32
    width, height: int32
    maze: Maze
    path: seq[Cell]

  Item* = object
    priority: int
    vertex: int

  MazeAStar* = ref object of Benchmark
    resultVal: uint32
    width, height: int32
    maze: Maze
    path: seq[Cell]

proc newCell(x, y: int): Cell
proc isWalkable*(kind: CellKind): bool
proc addNeighbor*(cell: Cell, neighbor: Cell)
proc reset*(cell: Cell)
proc newMaze*(width, height: int): Maze
proc updateNeighbors*(maze: Maze)
proc reset*(maze: Maze)
proc dig*(maze: Maze, startCell: Cell)
proc ensureOpenFinish*(maze: Maze, startCell: Cell)
proc generate*(maze: Maze)
proc middleCell*(maze: Maze): Cell
proc checksum*(maze: Maze): uint32
proc printToConsole*(maze: Maze)

proc newCell(x, y: int): Cell =
  new(result)
  result.kind = Wall
  result.neighbors = @[]
  result.x = x
  result.y = y

proc isWalkable*(kind: CellKind): bool =
  kind in {Space, Start, Finish}

proc addNeighbor*(cell: Cell, neighbor: Cell) =
  cell.neighbors.add(neighbor)

proc reset*(cell: Cell) =
  if cell.kind == Space:
    cell.kind = Wall

proc newMaze*(width, height: int): Maze =
  new(result)
  let w = max(width, 5)
  let h = max(height, 5)

  result.width = w
  result.height = h
  result.cells = newSeq[seq[Cell]](h)

  for y in 0..<h:
    result.cells[y] = newSeq[Cell](w)
    for x in 0..<w:
      result.cells[y][x] = newCell(x, y)

  result.start = result.cells[1][1]
  result.finish = result.cells[h-2][w-2]
  result.start.kind = Start
  result.finish.kind = Finish

  updateNeighbors(result)

proc updateNeighbors*(maze: Maze) =
  for y in 0..<maze.height:
    for x in 0..<maze.width:
      let cell = maze.cells[y][x]
      cell.neighbors.setLen(0)

      if x > 0 and y > 0 and x < maze.width - 1 and y < maze.height - 1:
        cell.addNeighbor(maze.cells[y-1][x])
        cell.addNeighbor(maze.cells[y+1][x])
        cell.addNeighbor(maze.cells[y][x+1])
        cell.addNeighbor(maze.cells[y][x-1])

        for _ in 0..<4:
          let i = nextInt(4)
          let j = nextInt(4)
          if i != j:
            swap(cell.neighbors[i], cell.neighbors[j])
      else:
        cell.kind = Border

proc reset*(maze: Maze) =
  for row in maze.cells:
    for cell in row:
      reset(cell)
  maze.start.kind = Start
  maze.finish.kind = Finish

proc dig*(maze: Maze, startCell: Cell) =
  var stack = newSeq[Cell]()
  stack.add(startCell)

  while stack.len > 0:
    let cell = stack.pop()

    var walkable = 0
    for n in cell.neighbors:
      if n.kind.isWalkable:
        inc walkable

    if walkable == 1:
      cell.kind = Space
      for n in cell.neighbors:
        if n.kind == Wall:
          stack.add(n)

proc ensureOpenFinish*(maze: Maze, startCell: Cell) =
  var stack = newSeq[Cell]()
  stack.add(startCell)

  while stack.len > 0:
    let cell = stack.pop()

    cell.kind = Space

    var walkable = 0
    for n in cell.neighbors:
      if n.kind.isWalkable:
        inc walkable

    if walkable <= 1:
      for n in cell.neighbors:
        if n.kind == Wall:
          stack.add(n)

proc generate*(maze: Maze) =
  for n in maze.start.neighbors:
    if n.kind == Wall:
      maze.dig(n)

  for n in maze.finish.neighbors:
    if n.kind == Wall:
      maze.ensureOpenFinish(n)

proc middleCell*(maze: Maze): Cell =
  maze.cells[maze.height div 2][maze.width div 2]

proc checksum*(maze: Maze): uint32 =
  var hasher = 2166136261'u32
  const prime = 16777619'u32

  for y in 0..<maze.height:
    for x in 0..<maze.width:
      if maze.cells[y][x].kind == Space:
        let val = uint32(x * y)
        hasher = (hasher xor val) * prime

  hasher

proc printToConsole*(maze: Maze) =
  for y in 0..<maze.height:
    for x in 0..<maze.width:
      case maze.cells[y][x].kind
      of Space: stdout.write(" ")
      of Wall: stdout.write("\e[34m#\e[0m")
      of Border: stdout.write("\e[31mO\e[0m")
      of Start: stdout.write("\e[32m>\e[0m")
      of Finish: stdout.write("\e[32m<\e[0m")
      of Path: stdout.write("\e[33m.\e[0m")
    echo ""
  echo ""

proc newMazeGenerator*(): Benchmark =
  MazeGenerator()

method name(self: MazeGenerator): string = "Maze::Generator"

method prepare(self: MazeGenerator) =
  self.width = int32(self.config_val("w"))
  self.height = int32(self.config_val("h"))
  self.maze = newMaze(self.width.int, self.height.int)
  self.resultVal = 0

method run(self: MazeGenerator, iteration_id: int) =
  reset(self.maze)
  generate(self.maze)
  self.resultVal += uint32(self.maze.middleCell().kind)

method checksum(self: MazeGenerator): uint32 =
  self.resultVal + self.maze.checksum()

registerBenchmark("Maze::Generator", newMazeGenerator)

proc newMazeBFS*(): Benchmark =
  MazeBFS()

method name(self: MazeBFS): string = "Maze::BFS"

method prepare(self: MazeBFS) =
  self.width = int32(self.config_val("w"))
  self.height = int32(self.config_val("h"))
  self.maze = newMaze(self.width.int, self.height.int)
  generate(self.maze)
  self.resultVal = 0
  self.path = @[]

proc bfs(maze: Maze, start, target: Cell): seq[Cell] =
  if start == target:
    return @[start]

  var queue = initDeque[int]()
  var visited = newSeq[seq[bool]](maze.height)
  for y in 0..<maze.height:
    visited[y] = newSeq[bool](maze.width)
    for x in 0..<maze.width:
      visited[y][x] = false

  var pathNodes = newSeq[PathNode]()

  visited[start.y][start.x] = true
  pathNodes.add(PathNode(cell: start, parent: -1))
  queue.addLast(0)

  while queue.len > 0:
    let pathId = queue.popFirst()
    let node = pathNodes[pathId]

    for neighbor in node.cell.neighbors:
      if neighbor == target:
        var cur = pathId
        var res = @[target]
        while cur >= 0:
          res.add(pathNodes[cur].cell)
          cur = pathNodes[cur].parent
        return res.reversed()

      if neighbor.kind.isWalkable and not visited[neighbor.y][neighbor.x]:
        visited[neighbor.y][neighbor.x] = true
        pathNodes.add(PathNode(cell: neighbor, parent: pathId))
        queue.addLast(pathNodes.len - 1)

  @[]

proc midCellChecksum(path: seq[Cell]): uint32 =
  if path.len == 0: return 0
  let cell = path[path.len div 2]
  uint32(cell.x * cell.y)

method run(self: MazeBFS, iteration_id: int) =
  self.path = bfs(self.maze, self.maze.start, self.maze.finish)
  self.resultVal += uint32(self.path.len)

method checksum(self: MazeBFS): uint32 =
  self.resultVal + midCellChecksum(self.path)

registerBenchmark("Maze::BFS", newMazeBFS)

proc `<`(a, b: Item): bool =
  if a.priority != b.priority:
    a.priority < b.priority
  else:
    a.vertex < b.vertex

proc newMazeAStar*(): Benchmark =
  MazeAStar()

method name(self: MazeAStar): string = "Maze::AStar"

method prepare(self: MazeAStar) =
  self.width = int32(self.config_val("w"))
  self.height = int32(self.config_val("h"))
  self.maze = newMaze(self.width.int, self.height.int)
  generate(self.maze)
  self.resultVal = 0
  self.path = @[]

proc heuristic(a, b: Cell): int =
  abs(a.x - b.x) + abs(a.y - b.y)

proc idx(y, x, width: int): int =
  y * width + x

proc astar(maze: Maze, start, target: Cell, width, height: int): seq[Cell] =
  if start == target:
    return @[start]

  let size = width * height
  var cameFrom = newSeq[int](size)
  var gScore = newSeq[int](size)
  var bestF = newSeq[int](size)

  for i in 0..<size:
    cameFrom[i] = -1
    gScore[i] = high(int)
    bestF[i] = high(int)

  let startIdx = idx(start.y, start.x, width)
  let targetIdx = idx(target.y, target.x, width)

  var openSet = initHeapQueue[Item]()
  var inOpen = newSeq[bool](size)

  gScore[startIdx] = 0
  let fStart = heuristic(start, target)
  openSet.push(Item(priority: fStart, vertex: startIdx))
  bestF[startIdx] = fStart
  inOpen[startIdx] = true

  while openSet.len > 0:
    let current = openSet.pop()
    let currentIdx = current.vertex
    inOpen[currentIdx] = false

    if currentIdx == targetIdx:
      var cur = currentIdx
      var res = newSeq[Cell]()
      while cur != -1:
        let y = cur div width
        let x = cur mod width
        res.add(maze.cells[y][x])
        cur = cameFrom[cur]
      return res.reversed()

    let currentY = currentIdx div width
    let currentX = currentIdx mod width
    let currentCell = maze.cells[currentY][currentX]
    let currentG = gScore[currentIdx]

    for neighbor in currentCell.neighbors:
      if not neighbor.kind.isWalkable:
        continue

      let neighborIdx = idx(neighbor.y, neighbor.x, width)
      let tentativeG = currentG + 1

      if tentativeG < gScore[neighborIdx]:
        cameFrom[neighborIdx] = currentIdx
        gScore[neighborIdx] = tentativeG
        let fNew = tentativeG + heuristic(neighbor, target)

        if fNew < bestF[neighborIdx]:
          bestF[neighborIdx] = fNew
          openSet.push(Item(priority: fNew, vertex: neighborIdx))
          inOpen[neighborIdx] = true

  @[]

method run(self: MazeAStar, iteration_id: int) =
  self.path = astar(self.maze, self.maze.start, self.maze.finish,
                    self.maze.width, self.maze.height)
  self.resultVal += uint32(self.path.len)

method checksum(self: MazeAStar): uint32 =
  self.resultVal + midCellChecksum(self.path)

registerBenchmark("Maze::AStar", newMazeAStar)
